src/main/java/com/mraof/minestuck/tileentity/redstone/StructureCoreTileEntity.java

```diff
@@ -0,0 +1,256 @@
+{
+	private int tickCycle;
+	@Nonnull
+	private ActionType actionType;
+	private int shutdownRange = 32;
+	private boolean hasWiped = false;
+	public final Structure<?>[] SOLVABLE_STRUCTURES = new Structure<?>[]{MSFeatures.FROG_TEMPLE};
+	
+	/**
+	 * READ AND WIPE checks to see if the structure piece hasBeenCompleted variable is true, and then prevents puzzle blocks not meant to be utilized for survival players from working
+	 * READ AND REDSTONE checks to see if the structure piece hasBeenCompleted variable is true, and then gives off a redstone signal if so
+	 * WRITE turns the structure piece hasBeenCompleted variable to true if it is not yet
+	 */
+	public enum ActionType
+	{
+		READ_AND_WIPE,
+		READ_AND_REDSTONE,
+		WRITE;
+		
+		public static ActionType fromInt(int ordinal) //converts int back into enum
+		{
+			for(ActionType type : ActionType.values())
+			{
+				if(type.ordinal() == ordinal)
+					return type;
+			}
+			throw new IllegalArgumentException("Invalid ordinal of " + ordinal + " for structure core action type!");
+		}
+		
+		public String getNameNoSpaces()
+		{
+			return name().replace('_', ' ');
+		}
+	}
+	
+	public StructureCoreTileEntity()
+	{
+		super(MSTileEntityTypes.STRUCTURE_CORE.get());
+		actionType = ActionType.READ_AND_REDSTONE;
+	}
+	
+	@Override
+	public void tick()
+	{
+		if(level == null || level.isClientSide)
+			return;
+		
+		int cycleRate = (hasWiped && actionType == ActionType.READ_AND_WIPE) ? 600 : 100;
+		if(tickCycle >= cycleRate)
+		{
+			sendUpdate();
+			tickCycle = 0;
+		}
+		tickCycle++;
+	}
+	
+	private void sendUpdate()
+	{
+		if(level != null && level.isAreaLoaded(getBlockPos(), 1) && getBlockState().getValue(StructureCoreBlock.ACTIVE))
+		{
+			StructureManager structureManager = ((ServerWorld) level).structureFeatureManager();
+			StructureStart<?> structureStart = getStructureStart(structureManager);
+			if(structureStart != null && structureStart.isValid())
+			{
+				Debug.debugf("meets all conditions, now doing action type function");
+				if(actionType == ActionType.WRITE)
+				{
+					writeToStructure(structureStart);
+				} else if(actionType == ActionType.READ_AND_WIPE)
+				{
+					wipeSlate();
+				} else if(actionType == ActionType.READ_AND_REDSTONE)
+				{
+					CoreCompatibleScatteredStructurePiece piece = getStructurePiece(structureStart);
+					
+					if(piece != null && piece.hasBeenCompleted())
+					{
+						if(!getBlockState().getValue(StructureCoreBlock.POWERED))
+							level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, true));
+					} else
+					{
+						level.setBlockAndUpdate(getBlockPos(), getBlockState().setValue(StructureCoreBlock.POWERED, false));
+					}
+				}
+			}
+		}
+	}
+	
+	public StructureStart<?> getStructureStart(StructureManager structureManager)
+	{
+		for(Structure<?> structureListIterate : SOLVABLE_STRUCTURES)
+		{
+			return structureManager.getStructureAt(getBlockPos(), true, structureListIterate);
```

Comment:
This'll just return for the first structure in `SOLVABLE_STRUCTURES`.
Also, have you considered configuring each individual structure core for a specific structure, instead of having a global list of affected structures?

Reply:
May take that consideration into effect at a later date

Reply:
Alright, but I'd still like some small fix for the immediate problem. Either:
- Use a single structure instead of a list of structures
- Fix the loop so that all entries in the list can get used (by returning the result of `getStructureAt()` only if it is not null)